File: backend/app/schemas/storage.py
"""Request/response models for the S3-compatible storage-targets API (2.2).

The access/secret keys are write-only: accepted on create/update, Fernet-encrypted
immediately, and NEVER echoed back — the read model exposes only a `keys_set`
boolean (CLAUDE.md rule 6).
"""

from datetime import datetime

from pydantic import BaseModel, Field

from app.models.storage import STORAGE_PROVIDERS, StorageTarget


class StorageTargetOut(BaseModel):
    """One storage target for the Settings list (no secret material)."""

    id: int
    name: str
    provider: str
    endpoint_url: str | None
    region: str | None
    bucket: str
    path_prefix: str | None
    use_ssl: bool
    enabled: bool
    keys_set: bool
    created_at: datetime
    updated_at: datetime

    @classmethod
    def from_model(cls, target: StorageTarget) -> "StorageTargetOut":
        return cls(
            id=target.id,
            name=target.name,
            provider=target.provider,
            endpoint_url=target.endpoint_url,
            region=target.region,
            bucket=target.bucket,
            path_prefix=target.path_prefix,
            use_ssl=target.use_ssl,
            enabled=target.enabled,
            keys_set=target.keys_set,
            created_at=target.created_at,
            updated_at=target.updated_at,
        )


class StorageTargetCreate(BaseModel):
    """Register a storage target. Both keys are required to create a usable one."""

    name: str = Field(min_length=1, max_length=120)
    provider: str = "aws"
    endpoint_url: str | None = Field(default=None, max_length=255)
    region: str | None = Field(default=None, max_length=64)
    bucket: str = Field(min_length=1, max_length=255)
    path_prefix: str | None = Field(default=None, max_length=255)
    access_key: str = Field(min_length=1, max_length=255)
    secret_key: str = Field(min_length=1, max_length=255)
    use_ssl: bool = True
    enabled: bool = True

    def provider_normalized(self) -> str:
        return self.provider if self.provider in STORAGE_PROVIDERS else "other"


class StorageTargetUpdate(BaseModel):
    """Patch a storage target. Keys are optional; a non-empty value re-encrypts,
    an empty string clears them, and omitting them leaves them unchanged."""

    name: str | None = Field(default=None, min_length=1, max_length=120)
    provider: str | None = None
    endpoint_url: str | None = Field(default=None, max_length=255)
    region: str | None = Field(default=None, max_length=64)
    bucket: str | None = Field(default=None, min_length=1, max_length=255)
    path_prefix: str | None = Field(default=None, max_length=255)
    access_key: str | None = Field(default=None, max_length=255)
    secret_key: str | None = Field(default=None, max_length=255)
    use_ssl: bool | None = None
    enabled: bool | None = None


class TestConnectionOut(BaseModel):
    """Result of a storage target's test-connection probe."""

    reachable: bool
    writable: bool
    latency_ms: int | None
    error: str | None = None

File: frontend/src/api/storage.ts
/**
 * Typed data layer for S3-compatible storage targets (session 2.2, CLAUDE.md
 * rule 9). A target is one bucket the platform pushes backup artifacts to for
 * offsite retention. The access/secret keys are write-only: sent on create/
 * update, Fernet-encrypted server-side, and NEVER returned — the read model
 * exposes only a `keys_set` boolean (rule 6).
 */

import { apiClient } from './client'

export type StorageProvider = 'aws' | 'minio' | 'backblaze' | 'wasabi' | 'other'

export interface StorageTarget {
  id: number
  name: string
  provider: string
  endpoint_url: string | null
  region: string | null
  bucket: string
  path_prefix: string | null
  use_ssl: boolean
  enabled: boolean
  /** Whether both credentials are set (the values themselves are never sent). */
  keys_set: boolean
  created_at: string
  updated_at: string
}

export interface StorageTargetCreate {
  name: string
  provider: StorageProvider
  endpoint_url?: string | null
  region?: string | null
  bucket: string
  path_prefix?: string | null
  access_key: string
  secret_key: string
  use_ssl: boolean
  enabled: boolean
}

/**
 * Patch a target. Keys are optional: a non-empty value re-encrypts, an empty
 * string clears them, and omitting them leaves the stored keys untouched.
 */
export interface StorageTargetUpdate {
  name?: string
  provider?: StorageProvider
  endpoint_url?: string | null
  region?: string | null
  bucket?: string
  path_prefix?: string | null
  access_key?: string
  secret_key?: string
  use_ssl?: boolean
  enabled?: boolean
}

export interface TestConnectionResult {
  reachable: boolean
  writable: boolean
  latency_ms: number | null
  error: string | null
}

export const storageApi = {
  list: () => apiClient.get<StorageTarget[]>('/api/storage-targets'),
  get: (id: number) => apiClient.get<StorageTarget>(`/api/storage-targets/${id}`),
  create: (payload: StorageTargetCreate) =>
    apiClient.post<StorageTarget>('/api/storage-targets', payload),
  update: (id: number, payload: StorageTargetUpdate) =>
    apiClient.patch<StorageTarget>(`/api/storage-targets/${id}`, payload),
  remove: (id: number) => apiClient.delete<void>(`/api/storage-targets/${id}`),
  testConnection: (id: number) =>
    apiClient.post<TestConnectionResult>(`/api/storage-targets/${id}/test-connection`),
}

export const STORAGE_PROVIDER_LABEL: Record<string, string> = {
  aws: 'AWS S3',
  minio: 'MinIO',
  backblaze: 'Backblaze B2',
  wasabi: 'Wasabi',
  other: 'S3-compatible',
}
